src/Plugins/Loader/AssemblyLoadContextBuilder.cs

```diff
@@ -128,7 +128,9 @@ public AssemblyLoadContextBuilder AddManagedLibrary(ManagedLibrary library)
         {
             ValidateRelativePath(library.AdditionalProbingPath);
 
-            _managedLibraries.Add(library.Name.Name, library);
+            if (!_managedLibraries.ContainsKey(library.Name.Name))
```

Comment:
Instead of ignoring this conflict, would it be better to replace? In the end, each loader can only have one version of a library by the same name. Where are you running into conflicts?

Reply:
If I run without my modifications to your project then my msbuild PowerShellTaskFactory will fail to load with, for example:

```
System.ArgumentException
  HResult=0x80070057
  Message=An item with the same key has already been added. Key: pwrshplugin
  Source=System.Private.CoreLib
  StackTrace:
   at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException[T](T key)
>	McMaster.NETCore.Plugins.dll!McMaster.NETCore.Plugins.Loader.AssemblyLoadContextBuilder.AddNativeLibrary(McMaster.NETCore.Plugins.LibraryModel.NativeLibrary library) Line 145	C#
 	McMaster.NETCore.Plugins.dll!McMaster.NETCore.Plugins.Loader.DependencyContextExtensions.AddDependencyContext(McMaster.NETCore.Plugins.Loader.AssemblyLoadContextBuilder builder, Microsoft.Extensions.DependencyModel.DependencyContext dependencyContext) Line 125	C#
 	McMaster.NETCore.Plugins.dll!McMaster.NETCore.Plugins.Loader.DependencyContextExtensions.AddDependencyContext(McMaster.NETCore.Plugins.Loader.AssemblyLoadContextBuilder builder, string depsFilePath) Line 55	C#
 	McMaster.NETCore.Plugins.dll!McMaster.NETCore.Plugins.Loader.DependencyContextExtensions.TryAddDependencyContext(McMaster.NETCore.Plugins.Loader.AssemblyLoadContextBuilder builder, string depsFilePath, out System.Exception error) Line 31	C#
 	McMaster.NETCore.Plugins.dll!McMaster.NETCore.Plugins.PluginLoader.CreateLoadContext(string baseDir, McMaster.NETCore.Plugins.PluginConfig config, System.Type[] sharedTypes) Line 98	C#
 	McMaster.NETCore.Plugins.dll!McMaster.NETCore.Plugins.PluginLoader.PluginLoader(McMaster.NETCore.Plugins.PluginConfig config, string baseDir, System.Type[] sharedTypes) Line 86	C#
 	McMaster.NETCore.Plugins.dll!McMaster.NETCore.Plugins.PluginLoader.CreateFromAssemblyFile(string assemblyFile, System.Type[] sharedTypes) Line 48	C#
 	IaaMsbuildTasks.dll!IaaMsbuildTasks.ContextIsolatedTaskFactory.Loader.get() Line 30	C#
 	IaaMsbuildTasks.dll!IaaMsbuildTasks.ContextIsolatedTaskFactory.InnerTaskFactoryType.get() Line 45	C#
 	IaaMsbuildTasks.dll!IaaMsbuildTasks.ContextIsolatedTaskFactory.Initialize(string taskName, System.Collections.Generic.IDictionary<string, Microsoft.Build.Framework.TaskPropertyInfo> parameterGroup, string taskBody, Microsoft.Build.Framework.IBuildEngine taskFactoryLoggingHost) Line 103	C#
```

I observe that this pwrshplugin assembly (which I am not familiar with but apparently is part of my nuget dependency graph) is present in the following locations of my plugins publish folder:

publish\runtimes\win-arm\native\pwrshplugin.dll
publish\runtimes\win-arm64\native\pwrshplugin.dll
publish\runtimes\win-x64\native\pwrshplugin.dll
publish\runtimes\win-x86\native\pwrshplugin.dll

Just now in debugging this to formulate this answer I realized that when I get the above exception that the value already present in the dictionary under the key pwrshplugin is for pwrshplugin.dll:

```
_nativeLibraries["pwrshplugin"] = 
    "microsoft.powershell.native\\6.0.4\\runtimes/win-x64/native/pwrshplugin.dll"
    AdditionalProbingPath: "microsoft.powershell.native\\6.0.4\\runtimes/win-x64/native/pwrshplugin.dll"
    AppLocalPath: "runtimes/win-x64/native/pwrshplugin.dll"
    Name: "pwrshplugin"
```
Whereas the 2nd value being added under the same dictionary key is actually the corresponding pwrshplugin.pdb file:

```
"pwrshplugin" = 
    "microsoft.powershell.native\\6.0.4\\runtimes/win-x64/native/pwrshplugin.pdb"
    AdditionalProbingPath: "microsoft.powershell.native\\6.0.4\\runtimes/win-x64/native/pwrshplugin.pdb"
    AppLocalPath: "runtimes/win-x64/native/pwrshplugin.pdb"
    Name: "pwrshplugin"
```
So my solution is pretty weak. I am sure that you can come up with a better one. I imagine the same problem is present with AddManagedLibrary() where I applied the same poor but-effective-for-me remedy.

Reply:
Ah, I see. Thanks for clarifying. I think there must be a bug in RID selection in DependencyContextExtensions. This would account for the issues in both native and rid-specific managed assets. I'll take a closer look at this tonight.